content/architecture/put-business-logic-into-services.md

```diff
@@ -33,12 +33,8 @@ We can refactor this and move the logic into a `PeopleService`.
   ...
 })
 export class SomeComponent implements OnInit, OnDestroy {
```

Comment:
Can you remove the OnInit hook implementation and import as well, please?

Reply:
I am not sure about this one, because I still prefer to use `ngOnInit` which makes testing this component easier too as you can control when it gets called. So would actually still keep the life-cycle-hook here, but move the http stuff and what not into a service to separate the business logic from the component.

Reply:
@d3lm when using Angular strict mode and explicitly setting the type of the variable, doing the following will throw an error:

```
@Component({
  ...
})
export class SomeComponent implements OnInit {
  data$: Observable<ItemModel[]>;

  constructor(private http: HttpClient) {}

  ngOnInit() {
    this.data$ = this.http.get('http://some-api.com/');
  }
}
```

As it would require you to also set the data$ type to  `Observable<ItemModel[]> | undefined`. So you need to assign the variable on declaration or in its constructor.

Reply:
@d3lm this is an example from the Ionic (Sidemenu) Starter Template:

```
import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

@Component({
  selector: 'app-folder',
  templateUrl: './folder.page.html',
  styleUrls: ['./folder.page.scss'],
})
export class FolderPage implements OnInit {
  public folder:string;

  constructor(private activatedRoute: ActivatedRoute) {}

  ngOnInit() {
    this.folder = this.activatedRoute.snapshot.paramMap.get('id');
  }
}
```

And I get the following errors:
![image](https://user-images.githubusercontent.com/5419161/140587538-d08bdc53-0e40-41d3-b554-43b4c1a92f37.png)

This one is related to the return value of the `paramMap.get` function
![image](https://user-images.githubusercontent.com/5419161/140587561-5dbc56a2-c57b-410b-8bdd-1c719cf1e60d.png)

Reply:
And, IIRC, Angular strict mode is enabled by default on new projects since v11 or v12. Should we add a new topic about strict mode?

Reply:
Yes that's true for strict mode. What I typically do in this case is to use the `!` operator because `ngOnInit` is guaranteed to be called and that life cycle hook gives you more control overall, whereas if you call the service directly as an assignment it gets called when the class is gets instantiated.

Example:

```ts
@Component({
  ...
})
export class SomeComponent implements OnInit {
  data$!: Observable<ItemModel[]>;

  constructor(private http: HttpClient) {}

  ngOnInit() {
    this.data$ = this.http.get('http://some-api.com/');
  }
}
```

Reply:
I am curious what other people say about this.

What do you think @geromegrignon?

Reply:
There were some discussions about that lately with the Angular team on Angular Discord server and they really encourage us to avoid using **!** as much as possible.
It makes sense for `@ViewChild`  or complex `@Input` which can't be initialized with a default value but otherwise, you should use inline initialization or the constructor.

@aikidave It might be a good addition to [Property binding best practices](https://angular.io/guide/property-binding-best-practices#property-binding-best-practices) documentation as it's a recurrent question on Discord: 
> Should I use inline, constructor, or ngOnInit initialization?

Adding the link to your related question on Twitter as a reference: https://twitter.com/elmd_/status/1456922225387491333

Reply:
I can't find the original discussion from Discord but IIRC the recommendations about  **!** usage were from @gkalpak

Reply:
In this specific case I would personally always make the `Observable` `readonly` which also would make it obligated to be initialized in the `constructor` or just after de declaration of the variable.

So to me there are 2 specific clear cases here. If the declaration of a variable does not obligate you to initialize it in the constructor I would always go to `ngOnInit` .. like @d3lm already stated, this helps with testing that logic as well.

Reply:
I almost never use `ngOnInit`. I know this was initial recommendation from Angular team, as TS added better types and that it doesn't make any meaningful difference in perf, this is what TypeSscript team at Google was recommending:
More general TypeScript rule: https://ts.dev/style/#field-initializers

It does make component more concise.

Even with `@Input` decorators I prefer setters.

As for testing, I recommend to test component through DOM interactions, not calling component methods directly. That means that wether init logic is happing in constructor or in `ngOnInit` is component's internal implementation details.

I welcome this change :)

Reply:
> I can't find the original discussion from Discord but IIRC the recommendations about **!** usage were from @gkalpak

@d3lm @geromegrignon Here's the link to the discussion (with alx) in the angular discord: https://discord.com/channels/748677963142135818/748806247926726736/836327684727636019